include/adaflo/sharp_interface_util.h

```diff
@@ -0,0 +1,1040 @@
+  GridGenerator::MarchingCubeAlgorithm<dim> mc(mapping,
+                                               dof_handler.get_fe(),
+                                               n_subdivisions);
+
+  AffineConstraints<double> constraints; // TODO: use the right ones
+
+  FEPointEvaluation<1, dim> phi_curvature(mapping, dof_handler.get_fe());
+
+  FESystem<dim>               fe_dim(dof_handler.get_fe(), dim);
+  FEPointEvaluation<dim, dim> phi_normal(mapping, fe_dim);
+  FEPointEvaluation<dim, dim> phi_force(mapping, dof_handler_dim.get_fe());
+
+  std::vector<double>                  buffer;
+  std::vector<double>                  buffer_dim;
+  std::vector<types::global_dof_index> local_dof_indices;
+
+  // loop over all cells
+  for (const auto i : dof_handler.get_triangulation().active_cell_iterators())
+    {
+      typename DoFHandler<dim>::active_cell_iterator cell = {
+        &dof_handler.get_triangulation(), i->level(), i->index(), &dof_handler};
+
+      typename DoFHandler<dim>::active_cell_iterator cell_dim = {
+        &dof_handler.get_triangulation(), i->level(), i->index(), &dof_handler_dim};
+
+      // determine if cell is cut by the interface and if yes, determine the quadrature
+      // point location and weight
+      const auto [points, weights] =
+        [&]() -> std::tuple<std::vector<Point<dim>>, std::vector<double>> {
+        // determine points and cells of aux surface triangulation
+        std::vector<Point<dim>>          vertices;
+        std::vector<::CellData<dim - 1>> cells;
+
+        // run square/cube marching algorithm
+        mc.process_cell(cell, ls_vector, vertices, cells);
+
+        if (vertices.size() == 0)
+          return {}; // cell is not cut by interface -> no quadrature points have the be
+                     // determined
+
+        std::vector<Point<dim>> points;
+        std::vector<double>     weights;
+
+        // create aux triangulation of subcells
+        Triangulation<dim - 1, dim> tria;
+        tria.create_triangulation(vertices, cells, {});
+
+        FE_Nothing<dim - 1, dim> fe;
+        FEValues<dim - 1, dim>   fe_eval(fe,
+                                       surface_quad,
+                                       update_quadrature_points | update_JxW_values);
+
+        // loop over all cells ...
+        for (const auto &cell : tria.active_cell_iterators())
+          {
+            fe_eval.reinit(cell);
+
+            // ... and collect quadrature points and weights
+            for (const auto q : fe_eval.quadrature_point_indices())
+              {
+                points.emplace_back(
+                  mapping.transform_real_to_unit_cell(i, fe_eval.quadrature_point(q)));
+                weights.emplace_back(fe_eval.JxW(q));
+              }
+          }
+        return {points, weights};
+      }();
+
+      if (points.size() == 0)
+        continue; // cell is not cut but the interface -> nothing to do
+
+      // proceed as usual
```

Comment:
@kronbichler A first implementation using MS/MC. The shown code uses the new algorithms to determine the quadrature points on the interface inside of a single cell (incl. JxW). After these lines, the code is as usual, using `FEPointEvaluation`.

The spurious-currents example is much more stable (the high velocities are still observable). However, the rising-bubble example still crashes after 4 time steps. I will investigate now.

Reply:
If we should proceed here, we probably let MS/MC compute directly the unit cell coordinate and compute `JxW` without creating a triangulation (which is probably the most expensive operation here).